### PR TITLE
Implement lifecycle tcp hook

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
> // TODO: implement a realistic TCP lifecycle hook

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/cc tallclair
 
**Release note**:

```release-note
Implement lifecycle tcp hook
```

### CODE REVIEW

File: pkg/kubelet/lifecycle/handlers.go [hunk 1]
```diff
@@ -113,21 +121,46 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
 		}
 		host = status.IP
 	}
-	var port int
-	if handler.HTTPGet.Port.Type == intstr.String && len(handler.HTTPGet.Port.StrVal) == 0 {
-		port = 80
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: why did you remove the defaulting?

File: pkg/kubelet/lifecycle/handlers.go [hunk 2]
```diff
@@ -113,21 +121,46 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+	port, err := resolvePort(handler.HTTPGet.Port, container)
+	if err != nil {
+		return "", err
 	}
 	url := fmt.Sprintf("http://%s/%s", net.JoinHostPort(host, strconv.Itoa(port)), handler.HTTPGet.Path)
 	resp, err := hr.httpGetter.Get(url)
 	return getHttpRespBody(resp), err
 }
 
+func (hr *HandlerRunner) runSocketHandler(pod *v1.Pod, container *v1.Container, handler *v1.Handler) (string, error) {
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: nit: call this `runTCPSocketHandler` in case we ever add other socket types.

File: pkg/kubelet/lifecycle/handlers.go [hunk 3]
```diff
@@ -113,21 +121,46 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+	host := handler.TCPSocket.Host
+	// default to pod ip
+	if len(host) == 0 {
+		status, err := hr.containerManager.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+		if err != nil {
+			glog.Errorf("Unable to get pod info, event handlers may be invalid.")
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: As a general rule, you should never log an error & return it. Just add this context & return the error (as you do below)

File: pkg/kubelet/lifecycle/handlers.go [hunk 4]
```diff
@@ -113,21 +121,46 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+			return "", err
+		}
+		if status.IP == "" {
+			return "", fmt.Errorf("failed to find networking container: %v", status)
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: this error isn't necessary accurate. Just say failed to get pod IP.

File: pkg/kubelet/lifecycle/handlers.go [hunk 5]
```diff
@@ -113,21 +121,46 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+		}
+		host = status.IP
+	}
+	port, err := resolvePort(handler.HTTPGet.Port, container)
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: s/HTTPGet/TCPSocket/

Please add a unit test that would catch this.

File: pkg/kubelet/lifecycle/handlers.go [hunk 6]
```diff
@@ -113,21 +121,46 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+	if err != nil {
+		return "", err
+	}
+
+	addr := net.JoinHostPort(host, strconv.Itoa(port))
+	conn, err := net.DialTimeout("tcp", addr, 10*time.Second)
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: nit: make the timeout a constant

Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: Can you deduplicate this with [DoTCPProbe](https://github.com/kubernetes/kubernetes/blob/152b0c12da40d85b087100b805cc348db613ee9d/pkg/probe/tcp/tcp.go#L47)

**Review by tallclair — COMMENTED** [review]
Please add unit tests

File: pkg/kubelet/lifecycle/handlers.go (hunk 1 above)
Comment (dixudx) [reply pkg/kubelet/lifecycle/handlers.go]: > why did you remove the defaulting?

Because this block of codes will never be run. Since apiserver will do port validation where `port=0` is invalid.

https://github.com/kubernetes/kubernetes/blob/0048d2da400b8c48ae83acc6a223a65f3551674a/pkg/apis/core/validation/validation.go#L2417-L2432

**Comment by liggitt** [comment]
/unassign

once this is reviewed and has lgtm, feel free to reassign it to me

File: pkg/kubelet/lifecycle/handlers.go [hunk 7]
```diff
@@ -69,6 +77,13 @@ func (hr *HandlerRunner) Run(containerID kubecontainer.ContainerID, pod *v1.Pod,
 			klog.V(1).Infof(msg)
 		}
 		return msg, err
+	case handler.TCPSocket != nil:
+		msg, err := hr.runTCPSocketHandler(pod, container, handler)
+		if err != nil {
+			msg = fmt.Sprintf("Http lifecycle hook (%s) for Container %q in Pod %q failed - error: %v, message: %q", handler.TCPSocket.String(), container.Name, format.Pod(pod), err, msg)
```
Comment (peter-wangxu) [inline pkg/kubelet/lifecycle/handlers.go]: Http -> Tcp

**Review by peter-wangxu — COMMENTED** [review]
Just nit:)

File: pkg/kubelet/lifecycle/handlers.go (hunk 7 above)
Comment (dixudx) [reply pkg/kubelet/lifecycle/handlers.go]: @peter-wangxu Great catch.

**Review by tallclair — COMMENTED** [review]
Sorry, I just realized I left a bunch of comments, but never submitted the review...

File: pkg/kubelet/lifecycle/handlers.go (hunk 1 above)
Comment (dixudx) [reply pkg/kubelet/lifecycle/handlers.go]: Oh, right. Forgot static pods.

**Comment by dixudx** [comment]
@tallclair  Updated. PTAL. Thanks.

File: pkg/kubelet/lifecycle/handlers.go [hunk 8]
```diff
@@ -30,24 +31,31 @@ import (
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
 	"k8s.io/kubernetes/pkg/kubelet/util/format"
+	"k8s.io/kubernetes/pkg/probe/tcp"
 	"k8s.io/kubernetes/pkg/security/apparmor"
 )
 
+const (
+	TCPSocketTimeOut = 1 * time.Second
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: How did you come up with 1 second? I'm not saying that 1 second or 10 seconds is wrong, I'd just like to know that this isn't an arbitrary made up number.

Comment (dixudx) [reply pkg/kubelet/lifecycle/handlers.go]: Actually the default timeout to setup a tcp connection is 127 seconds. I think that's too long for a pod probe.

**Comment by dixudx** [comment]
Rebased. @tallclair @yujuhong PTAL. Thanks.
